crates/services/sync/src/import/cache.rs

```diff
@@ -0,0 +1,533 @@
+        RangeInclusive,
+    },
+};
+
+use fuel_core_services::SharedMutex;
+use fuel_core_types::blockchain::{
+    SealedBlock,
+    SealedBlockHeader,
+};
+
+use super::Batch;
+
+/// The cache that stores the fetched headers and blocks.
+#[derive(Clone, Debug)]
+pub struct Cache(SharedMutex<BTreeMap<u32, CachedData>>);
+
+#[derive(Debug, Clone)]
+#[allow(clippy::large_enum_variant)]
+pub enum CachedData {
+    Header(SealedBlockHeader),
+    Block(SealedBlock),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum CachedDataBatch {
+    Headers(Batch<SealedBlockHeader>),
+    Blocks(Batch<SealedBlock>),
+    None(Range<u32>),
+}
+
+impl CachedDataBatch {
+    pub fn is_range_empty(&self) -> bool {
+        match self {
+            CachedDataBatch::None(range) => range.is_empty(),
+            CachedDataBatch::Blocks(batch) => batch.range.is_empty(),
+            CachedDataBatch::Headers(batch) => batch.range.is_empty(),
+        }
+    }
+}
+
+impl Cache {
+    pub fn new() -> Self {
+        Self(SharedMutex::new(BTreeMap::new()))
+    }
+
+    pub fn insert_blocks(&mut self, batch: Batch<SealedBlock>) {
+        let mut lock = self.0.lock();
+        for block in batch.results {
+            lock.insert(**block.entity.header().height(), CachedData::Block(block));
+        }
+    }
+
+    pub fn insert_headers(&mut self, batch: Batch<SealedBlockHeader>) {
+        let mut lock = self.0.lock();
+        for header in batch.results {
+            lock.insert(**header.entity.height(), CachedData::Header(header));
+        }
+    }
+
+    pub fn get_chunks(
+        &self,
+        range: RangeInclusive<u32>,
+        max_chunk_size: usize,
+    ) -> futures::stream::Iter<std::vec::IntoIter<CachedDataBatch>> {
+        let end = (*range.end()).saturating_add(1);
+        let chunk_size_u32 = u32::try_from(max_chunk_size)
+            .expect("The size of the chunk can't exceed `u32`");
+
+        let cache_iter = self.collect_cache_data(range.clone());
+        let mut current_height = *range.start();
+        let mut chunks = Vec::new();
+        let mut current_chunk = CachedDataBatch::None(0..0);
+
+        for (height, data) in cache_iter {
+            if height == current_height {
+                current_chunk =
+                    self.handle_current_chunk(current_chunk, data, height, &mut chunks);
+                current_chunk = self.check_chunk_limit(
+                    current_chunk,
+                    max_chunk_size,
+                    &mut chunks,
+                    current_height,
+                );
+            } else {
+                self.push_current_chunk(&mut chunks, current_chunk);
+                self.push_missing_chunks(
+                    &mut chunks,
+                    current_height,
+                    height,
+                    max_chunk_size,
+                    chunk_size_u32,
+                    end,
+                );
+                current_chunk = self.prepare_next_chunk(data, height);
```

Comment:
I applied you logic but changed `height = current_height` to `current_height = height` and it work and I think that what you meant because `height` is immutable and represent the last cache object we found and it's managed by the iterator